cpp/velox/compute/VeloxInitializer.cc

```diff
@@ -0,0 +1,191 @@
+#endif
+
+  auto properties = std::make_shared<const velox::core::MemConfig>(configurationValues);
+  auto hiveConnector =
+      velox::connector::getConnectorFactory(velox::connector::hive::HiveConnectorFactory::kHiveConnectorName)
+          ->newConnector(kHiveConnectorId, properties, ioExecutor_.get());
+
+  registerConnector(hiveConnector);
+  velox::parquet::registerParquetReaderFactory(velox::parquet::ParquetReaderType::NATIVE);
+  velox::dwrf::registerDwrfReaderFactory();
+  // Register Velox functions
+  registerAllFunctions();
+
+  InitCache(conf);
+}
+
+velox::memory::MemoryAllocator* VeloxInitializer::getAsyncDataCache() {
+  if (asyncDataCache_ != nullptr) {
+    return asyncDataCache_.get();
+  }
+  return velox::memory::MemoryAllocator::getInstance();
+}
+
+void VeloxInitializer::InitCache(std::unordered_map<std::string, std::string>& conf) {
+  auto key = conf.find(kVeloxCacheEnabled);
+  if (key != conf.end() && boost::algorithm::to_lower_copy(conf[kVeloxCacheEnabled]) == "true") {
+    FLAGS_ssd_odirect = true;
+    if (conf.find(kVeloxSsdODirectEnabled) != conf.end() &&
+        boost::algorithm::to_lower_copy(conf[kVeloxSsdODirectEnabled]) == "false") {
+      FLAGS_ssd_odirect = false;
+    }
+    uint64_t memCacheSize = std::stol(kVeloxMemCacheSizeDefault);
+    uint64_t ssdCacheSize = std::stol(kVeloxSsdCacheSizeDefault);
+    int32_t ssdCacheShards = std::stoi(kVeloxSsdCacheShardsDefault);
+    int32_t ssdCacheIOThreads = std::stoi(kVeloxSsdCacheIOThreadsDefault);
+    int32_t ioThreads = std::stoi(kVeloxIOThreadsDefault);
+    std::string ssdCachePathPrefix = kVeloxSsdCachePathDefault;
+    for (auto& [k, v] : conf) {
+      if (k == kVeloxMemCacheSize)
+        memCacheSize = std::stol(v);
+      if (k == kVeloxSsdCacheSize)
+        ssdCacheSize = std::stol(v);
+      if (k == kVeloxSsdCacheShards)
+        ssdCacheShards = std::stoi(v);
+      if (k == kVeloxSsdCachePath)
+        ssdCachePathPrefix = v;
+      if (k == kVeloxIOThreads)
+        ioThreads = std::stoi(v);
+      if (k == kVeloxSsdCacheIOThreads)
+        ssdCacheIOThreads = std::stoi(v);
+    }
+    std::string ssdCachePath = ssdCachePathPrefix + "/cache." + genUuid() + ".";
+    ssdCacheExecutor_ = std::make_unique<folly::IOThreadPoolExecutor>(ssdCacheIOThreads);
+    ioExecutor_ = std::make_unique<folly::IOThreadPoolExecutor>(ioThreads);
+    auto ssd =
+        std::make_unique<velox::cache::SsdCache>(ssdCachePath, ssdCacheSize, ssdCacheShards, ssdCacheExecutor_.get());
+
+    std::error_code ec;
+    const std::filesystem::space_info si = std::filesystem::space(ssdCachePathPrefix, ec);
+    if (si.available < ssdCacheSize) {
+      VELOX_FAIL(
+          "not enough space for ssd cache in " + ssdCachePath + " cache size: " + std::to_string(ssdCacheSize) +
+          "free space: " + std::to_string(si.available))
+    }
+
+    velox::memory::MmapAllocator::Options options;
+    options.capacity = memCacheSize;
```

Comment:
this line will place a limit on the capacity of allocator, so it should also count in `offheap.size`?

Reply:
This part should be counted in offheap.size and spark's storage memory.
@zhztheplayer Do we have interface to report native memory allocation to spark's storage memory?

But the mmapallocator isn't been counted yet,. So let's skip this part for now. It should be counted in executor.overhead now.

Reply:
> Do we have interface to report native memory allocation to spark's storage memory?

Not yet, Spark's storage memory API is only for its internal use for now. We can hack that API or just use execution memory without a spiller.

Is the cache going to be reused across Spark tasks? Execution memory is not for that case.

Reply:
@zhouyuan can the cache be shared in all task thread in the same executor? I'd think so.

@zhztheplayer do we hack the execution memory API?

Reply:
> @zhztheplayer do we hack the execution memory API?

Not it's a public API. We just used it.

Reply:
Can we submit a PR to Spark community to make storage API public?

Reply:
I am not sure, probably Spark has its reasons to make it private. Let's take more look into this part to see what we can do.